### PR TITLE
Added pressing and deleting with the [DELETE] key

### PR DESCRIPTION
**Pull request by SergeyG22**
added pressing and deleting with the [DELETE] key (issue #1356 [DELETE] key doesn't work in map editor)

### CODE REVIEW

File: mapeditor/mainwindow.cpp [hunk 1]
```diff
@@ -1181,3 +1181,13 @@ void MainWindow::on_actionPaste_triggered()
 	}
 }
 
+void MainWindow::keyPressEvent(QKeyEvent* event)
```
Comment (kambala-decapitator) [inline mapeditor/mainwindow.cpp:1184]: according to our current [coding guidelines](https://wiki.vcmi.eu/Coding_guidelines#Where_to_put_spaces) should be:
```suggestion
void MainWindow::keyPressEvent(QKeyEvent * event)
```

File: mapeditor/mainwindow.h [hunk 2]
```diff
@@ -161,4 +161,6 @@ public slots:
 
 	// command line options
 	QString mapFilePath;			// FilePath to the H3 or VCMI map to open
+protected:
+	virtual void keyPressEvent(QKeyEvent* event);
```
Comment (kambala-decapitator) [inline mapeditor/mainwindow.h:165]:
```suggestion

protected:
	virtual void keyPressEvent(QKeyEvent * event);
```

**Review by kambala-decapitator — CHANGES_REQUESTED** [review]
the approach itself looks good

**Comment by Nordsoft91** [comment]
Seems like delete and backspace are handled differently. Is it possible to handle them in same way?
Also, is it possible to define hotkey for button as it done for all other actions?

**Comment by SergeyG22** [comment]
'delete' and 'backspace' are handled differently because calling 'backspace' in a standard event loop doesn't work properly. Perhaps installing an event filter will solve this. But I do not know how to do it. The easiest way I've found is to use a hotkey.

**Comment by kambala-decapitator** [comment]
> 'delete' and 'backspace' are handled differently because calling 'backspace' in a standard event loop doesn't work properly. 

at least on macOS I don't see any issues with the following test code:

```cpp
#include <QtWidgets>

class MainWindow : public QMainWindow
{
protected:
	virtual void keyPressEvent(QKeyEvent* e) override
	{
		switch (e->key())
		{
		case Qt::Key_Backspace:
			qDebug("Key_Backspace");
			break;
		default:
			QMainWindow::keyPressEvent(e);
			break;
		}
	}
};

int main(int argc, char* argv[])
{
	QApplication a{argc, argv};
	MainWindow w;
	w.show();
	return a.exec();
}
```

> Perhaps installing an event filter will solve this.

`keyPressEvent` method is identical to handling key presses in event filter

> please handle Backspace key on macOS instead

my original intention was to handle Backspace on macOS and Delete on other platforms (`#ifdef` would've been the easiest solution), sorry for not being clear. I'd suggest to just do this removing all the complexity of the last commit.

**Comment by SergeyG22** [comment]
I don't have a macOS distribution. What you have given works well in the test project. Try adding a 'delete' handler to this code there as well. Still working? And now do the same in the map designer. The bundle of delete and backspace does not work on Windows 10. It seems that backspace is in conflict with something. I don't know anything about Mac. It's not just that I took out the backspace in the hotkey.

**Comment by kambala-decapitator** [comment]
I'll test how Backspace behaves in the editor on macOS

**Comment by kambala-decapitator** [comment]
@SergeyG22 have you tested erase functionality? For me it doesn't work at all on macOS - tried from toolbar, from menu, from instrument palette, with the new shortcut... @Nordsoft91 does it work for you?

btw you can attach key sequence directly to the action, no need to add a separate one or handle key event on main window. I also see that current shortcut is "press backspace, then press delete" which is obviously wrong.

**Comment by kambala-decapitator** [comment]
@SergeyG22 so, does erase work for you on your OS?

@Nordsoft91 and what about you on macOS?

**Comment by SergeyG22** [comment]
bandicam 2023-01-18 11-41-57-868

**Comment by kambala-decapitator** [comment]
thanks, I'm stupid: I was trying to erase land instead of objects =)

> btw you can attach key sequence directly to the action, no need to add a separate one or handle key event on main window. I also see that current shortcut is "press backspace, then press delete" which is obviously wrong.

so let's do this:
1. remove all your changes
2. remove shortcut of erase action from the .ui file
3. in code (in constructor) give the action Backspace key sequence on macOS and Delete on other platforms using `#ifdef`

**Comment by SergeyG22** [comment]
I'll do it as I have free time

**Review by Nordsoft91 — APPROVED** [review]
We can keep as is

**Comment by kambala-decapitator** [comment]
> We can keep as is

don't see any reason to bloat code with strange workarounds when doing things right is trivial

**Comment by Nordsoft91** [comment]
Fixed in other PR

**Pull request closed without merging**